crypto/cmp/cmp_ctx.c

```diff
@@ -0,0 +1,1143 @@
+    ASN1_OCTET_STRING_free(ctx->recipNonce);
+    sk_OSSL_CMP_ITAV_pop_free(ctx->geninfo_itavs, OSSL_CMP_ITAV_free);
+    sk_X509_pop_free(ctx->extraCertsOut, X509_free);
+
+    EVP_PKEY_free(ctx->newPkey);
+    X509_NAME_free(ctx->issuer);
+    X509_NAME_free(ctx->subjectName);
+    sk_GENERAL_NAME_pop_free(ctx->subjectAltNames, GENERAL_NAME_free);
+    sk_X509_EXTENSION_pop_free(ctx->reqExtensions, X509_EXTENSION_free);
+    sk_POLICYINFO_pop_free(ctx->policies, POLICYINFO_free);
+    X509_free(ctx->oldClCert);
+    X509_REQ_free(ctx->p10CSR);
+
+    sk_OSSL_CMP_ITAV_pop_free(ctx->genm_itavs, OSSL_CMP_ITAV_free);
+
+    sk_ASN1_UTF8STRING_pop_free(ctx->lastStatusString, ASN1_UTF8STRING_free);
+    X509_free(ctx->newClCert);
+    sk_X509_pop_free(ctx->caPubs, X509_free);
+    sk_X509_pop_free(ctx->extraCertsIn, X509_free);
+
+    OPENSSL_free(ctx);
+}
+
+/*
+ * returns the PKIStatus from the last CertRepMessage
+ * or Revocation Response, -1 on error
+ */
+int OSSL_CMP_CTX_get_status(OSSL_CMP_CTX *ctx)
+{
+    return ctx != NULL ? ctx->lastPKIStatus : -1;
+}
+
+/*
+ * returns the statusString from the last CertRepMessage
+ * or Revocation Response, NULL on error
+ */
+OSSL_CMP_PKIFREETEXT *OSSL_CMP_CTX_get0_statusString(OSSL_CMP_CTX *ctx)
+{
+    return ctx != NULL ? ctx->lastStatusString : NULL;
+}
+
+int CMP_CTX_set0_statusString(OSSL_CMP_CTX *ctx, OSSL_CMP_PKIFREETEXT *text)
+{
+    if (ctx == NULL)
+        return 0;
+
+    sk_ASN1_UTF8STRING_pop_free(ctx->lastStatusString, ASN1_UTF8STRING_free);
+    ctx->lastStatusString = text;
+    return 1;
+}
+
+int CMP_CTX_set0_validatedSrvCert(OSSL_CMP_CTX *ctx, X509 *cert)
+{
+    if (ctx == NULL)
+        return 0;
+
+    X509_free(ctx->validatedSrvCert);
+    ctx->validatedSrvCert = cert;
+    return 1;
+}
+
+/*
+ * Set callback function for checking if the cert is ok or should
+ * it be rejected.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set_certConf_cb(OSSL_CMP_CTX *ctx, OSSL_cmp_certConf_cb_t cb)
+{
+    if (ctx == NULL)
+        return 0;
+    ctx->certConf_cb = cb;
+    return 1;
+}
+
+/*
+ * Set argument, respectively a pointer to a structure containing arguments,
+ * optionally to be used by the certConf callback
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set_certConf_cb_arg(OSSL_CMP_CTX *ctx, void *arg)
+{
+    if (ctx == NULL)
+        return 0;
+    ctx->certConf_cb_arg = arg;
+    return 1;
+}
+
+/*
+ * Get argument, respectively the pointer to a structure containing arguments,
+ * optionally to be used by certConf callback
+ * returns callback argument set previously (NULL if not set or on error)
+ */
+void *OSSL_CMP_CTX_get_certConf_cb_arg(OSSL_CMP_CTX *ctx)
+{
+    if (ctx == NULL)
+        return NULL;
+    return ctx->certConf_cb_arg;
+}
+
+/*
+ * Set a callback function for log messages.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set_log_cb(OSSL_CMP_CTX *ctx, OSSL_cmp_log_cb_t cb)
```

Comment:
The upshot of our exchange on this topic was that the trace API unfortunately is not very usable for outputting anything except for (detailed) debugging info because by default it is disabled.

Error messages can of course be thrown in the usual way (using the error queue)
while for other use cases such as warnings and progress information we need to resort to some other solutions such as printing to `bio_out` or `bio_err` as will be done in `apps/cmp.c`.
The CMP app meanwhile emits a warning that the CMP lib can only produced progress info if the trace API is enabled.

The above CMP-specific log callback serves an orthogonal purpose, as required by our customers:
Regardless whether some event is signaled via the error queue or the trace API etc., the application should be to catch all diagnostic output, including source location details, and be redirect it to wherever it pleases.